### DISCUSSION TITLE
[Feature Request] Provide more triggers on which plugins may be loaded

### DISCUSSION THREAD

**Discussion by clorl** [alangrainger/obsidian-lazy-plugins#27] [Feature requests]
# Feature Description
Currently, there are a few triggers to choose when to load a plugin: Instant, Short Delay, Long Delay, Disabled. All of them are time-based.

You could implement more trigger types base on Obsidian provided events after which plugins may lazy-load, similarly to neovim's [lazy](https://lazy.folke.io/spec#spec-lazy-loading) package manager.

For example, a user could set a plugin to load when a specific file is opened, when any file is opened, when entering edit mode, etc.

Here are some examples of triggers that could be implemented : 
- AfterPlugin : Load the plugin after another plugin has loaded, useful if multiple plugins are used together or if they depend on another
- FileOpened : When any file is opened, if the plugin isn't loaded then load it. Alternatives could include specifying a filetype, directory, tag, regex for the title etc.
- FileCreated
- FileDeleted
- FileMoved
- EditStarted : When editing a page
- CustomAction : User could specify "groups" that they manually activate with a command. Eg, grouping all the plugins only useful when editing pages could be requested by the user when they need to.
- AppReady : Instead of having a "short" delay, request the plugin to load once Obsidian has loaded the editor
etc.

**Comment by codey-badger** [comment]
If there's a roadmap I would like to see this on it.

**Comment by kylek29** [comment]
I could see this as a very useful expansion. I do this manually to control the memory usage on Android -- very important on A15/A16 because of its aggressive memory manager (it evicts rather quickly).

To give a use-case, I have a custom JS script that looks for special property flags when a note loads, these kick off a loading group which enables plugins within that group.

For an example, I may have a "My Movie Reviews Table.md" note that has a property of `required_plugins: ['DataView', 'Meta-Bind', '...'] .. the customJS script reads that property on note load and then kicks off the required plugins if necessary. 


So one expansion is to allow each plugin to be added to custom groups (a plugin could belong to numerous different groups) the user creates and then have user-commands for enabling/disabling those groups.
